### PR TITLE
Add basic implementation for `system_context`

### PR DESCRIPTION
**Pull request by lucteo**
This adds the basic implementation of `system_context` according to P2079R3.

Making Lee's implementation work in stdexec, adding a few extra tests.

### CODE REVIEW

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:246]:
```suggestion
stdexec::__result_of<__exec_pool_operation_state, __exec_system_bulk_operation_state_impl*, __exec_pool_sender_t, __exec_system_bulk_shape, __exec_system_bulk_function_object>
```

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:40]: we should decide if we're ugly-ifying identifiers in this header or not and stick to it. I would prefer all identifiers to be ugly-ified.

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:80]: Are phases 2 and 3 planned for this PR, or will they happen in a future PR?

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:179]: there are a slew of `new`'s and no `delete`s or `unique_ptr`s. are all the allocations intended to leak?

**Review by ericniebler — CHANGES_REQUESTED** [review]
i'm excited for this. is this PR still evolving (phase 2, phase 3 as in the comments)? also, i'd like a plan about adding queries without breaking ABI.

and per request from @sean-parent, what is the plan for integrating this with system event loops? imagine a single-threaded embedded system running an event loop. how does that interoperate with this?

finally, i'd like to add phase 4: back this with windows thread pool on windows, or io_uring (?) on Linux, or GCD on MacOS. :-)))

**Comment by lucteo** [comment]
> i'm excited for this. is this PR still evolving (phase 2, phase 3 as in the comments)? also, i'd like a plan about adding queries without breaking ABI.
> 
> and per request from @sean-parent, what is the plan for integrating this with system event loops? imagine a single-threaded embedded system running an event loop. how does that interoperate with this?
> 
> finally, i'd like to add phase 4: back this with windows thread pool on windows, or io_uring (?) on Linux, or GCD on MacOS. :-)))

My plan for this PR was just to push what Lee had already done + some fixes + adding more tests to cover the requirements from the paper. Phases 2+ will be coming in following PRs

File: include/exec/system_scheduler.hpp
Comment (lucteo) [reply include/exec/system_scheduler.hpp:80]: In a future PR. I'll try to keep the PRs small and focused, and this one is already pretty big.

File: include/exec/system_scheduler.hpp
Comment (lucteo) [reply include/exec/system_scheduler.hpp:179]: I would mark all these as things still left to be done. The implementation is somehow between Phase 1 and Phase 2 (using Lee's plan).

File: include/exec/system_scheduler.hpp
Comment (lucteo) [reply include/exec/system_scheduler.hpp:328]: If that's ok with you, I'll fix all these TODOs in later PRs

File: include/exec/system_scheduler.hpp
Comment (lucteo) [reply include/exec/system_scheduler.hpp:443]: `set_error` is not promised by the paper. There was an error here and I removed this case. It seems that I should have removed other occurrences of `set_error`; will remove them now.

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [reply include/exec/system_scheduler.hpp:179]: I'm fine leaving phases 2 and 3 to later. i'm not fine with known memory leaks in `main`.

File: include/exec/system_scheduler.hpp
Comment (LeeHowes) [reply include/exec/system_scheduler.hpp:442]: I forget the deal with set_error in the paper. I think I just didn't have a good reason to put it here, but the expectation would be that a receiver would be capable of receiving errors. Whether we want to define some specific potential failure types or not I don't know.

File: include/exec/system_scheduler.hpp
Comment (LeeHowes) [reply include/exec/system_scheduler.hpp:262]: My implementation was initial and hacky, I wasn't aiming to follow the latest design guidelines.

**Comment by LeeHowes** [comment]
Yes the point I left this at was really early. As I wasn't writing it in the main repo I felt it was less work to leave memory leaks in and fix the APIs properly, rather than to make something correct but suboptimal first and evolve that to optimal correctness. The constraints may be different in stdexec.

**Comment by LeeHowes** [comment]
> i'm excited for this. is this PR still evolving (phase 2, phase 3 as in the comments)? also, i'd like a plan about adding queries without breaking ABI.
> 
> and per request from @sean-parent, what is the plan for integrating this with system event loops? imagine a single-threaded embedded system running an event loop. how does that interoperate with this?

This should fall out of the replacement mechanism. We swap out the entire implementation, with no API changes, with an event loop. Driving of the event loop is something you do directly on the implementation because it's very specific, and not
exposed through this API. 

Now, even loop in the sense of supporting timers etc is slightly different, and I think there we need to be clear that we can evolve the ABI safely behind the plugin mechanism. This is where I lost confidence in my ability to spend the time on it to get it right.

> finally, i'd like to add phase 4: back this with windows thread pool on windows, or io_uring (?) on Linux, or GCD on MacOS. :-)))

Yes, same applies. I don't think that is at all complicated from the POV of the paper. Once we have a plugin API, we just hide all that inside. Demonstrating that we can do so though is important. It can be just a test case in the repo.

**Comment by sean-parent** [comment]
If this is to support P2079R3, then I'd like to see an implementation layered earlier on some of the possible system thread pools. There is so much in that proposal that is not specified. What requirements are placed on the system scheduler? Most system schedulers support priorities and are elastic (i.e., blocking in the system thread pool will spin up additional threads to some limit).

> If the underlying system_context is unable to make progress on work created through system_scheduler instances, and the sender retrieved from the scheduler is connected to a receiver that supports the get_delegatee_scheduler query, work may scheduled on the scheduler returned by get_delegatee_scheduler at the time of the call to start, or at any later point before the work completes.

If I understand correctly, this is saying tasks can be stolen and executed on the current thread when work is scheduled. This may mean that work expected to happen concurrently on another thread is executed in the same thread context. In practice, this is a recipe for deadlocks on odd reentrancy bugs of code holding a lock while scheduling work. I did this in the portable STLab scheduler and had to pull it out and instead make the pool elastic. Unfortunately, the answer to "don't hold a mutex while scheduling work" wasn't viable in a large existing code base.

This approach also caused "Aw Snaps" on a web app where tasks intended to be scheduled async would get stolen in the main js run loop and exceed the browser time limit before returning to the main thread.

There is a good overview of the Window thread pool design [here](https://youtu.be/Z86b3Rd09sE?si=U4095X-OISWqejAj). Apple's GCD is a very similar model.

File: include/exec/system_scheduler.hpp
Comment (lucteo) [reply include/exec/system_scheduler.hpp:442]: After thinking some more on this, I think we need to have `set_error` here. Otherwise, how would the the implementation trigger some runtime errors (lack of memory, failure to allocate resources, etc.)

@ericniebler : at some point, long long time ago, there was a push to remove requirements for needing `set_error`. I take it that we do not want in P2300 to remove `std_error` as a requirement for any receiver, right?

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:28]:
```suggestion
// Virtual interfaces to underlying implementations for initial simplicity
```

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:102]: future wishlist: support compiling without RTTI

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:119]: These should be defined in situ to remain hidden friends.

```suggestion
  void set_value() noexcept;
  void set_stopped() noexcept;
  void set_error(std::exception_ptr) noexcept;

  friend void tag_invoke(stdexec::set_value_t, __exec_system_pool_receiver&& __self) noexcept {
    __self.set_value();
  }

  friend void tag_invoke(stdexec::set_stopped_t, __exec_system_pool_receiver&& __self) noexcept {
    __self.set_stopped();
  }

  friend void
    tag_invoke(stdexec::set_error_t, __exec_system_pool_receiver&& __self, std::exception_ptr __ptr) noexcept {
      __self.set_error(std::move(__ptr));
    }
```

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:153]:
```suggestion
inline void __exec_system_pool_receiver::set_value() noexcept {
```

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:158]:
```suggestion
inline void __exec_system_pool_receiver::set_stopped() noexcept {
```

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:166]:
```suggestion
inline void __exec_system_pool_receiver::set_error(
  std::exception_ptr __ptr) noexcept {
```

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [inline include/exec/system_scheduler.hpp:205]: Similarly define these in situ.

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [reply include/exec/system_scheduler.hpp:340]: i don't want to merge to main any symbols in the global namespace.

File: include/exec/system_scheduler.hpp
Comment (ericniebler) [reply include/exec/system_scheduler.hpp:461]: there is no longer a requirement for a receiver to have a set_error.

**Review by ericniebler — CHANGES_REQUESTED** [review]
it's ok for the other changes to wait, but the identifiers in the global namespace and the memory leaks need to be fixed before i can feel good about merging this.

**Comment by ericniebler** [comment]
The nvc++ test failure is odd:

```
test 607
        Start 607: a copied scheduler is equal to the original

607: Test command: /home/coder/stdexec/build/test/test.stdexec "a copied scheduler is equal to the original"
607: Working Directory: /home/coder/stdexec/build/test
607: Test timeout computed to be: 60
607: Filters: a copied scheduler is equal to the original
607: ===============================================================================
607: All tests passed (1 assertion in 1 test case)
607: 
607: pure virtual method called
607: terminate called without an active exception
607/715 Test #607: a copied scheduler is equal to the original ...................................................................................Subprocess aborted***Exception:   0.27 sec
```

The runtime failure is "`pure virtual method called`", but there are no `virtual` functions in the implementation. @lucteo can you think of anything that would cause this?

File: include/exec/system_context.hpp
Comment (ericniebler) [inline include/exec/system_context.hpp:74]:
```suggestion
    bool operator==(const system_scheduler&) const noexcept = default;
```

File: include/exec/system_context.hpp
Comment (ericniebler) [inline include/exec/system_context.hpp:98]: We need to switch to the new algorithm customization scheme, giving the `system_scheduler` a domain and customizing `bulk` there.

**Review by ericniebler — COMMENTED** [review]
This is looking very good.

File: include/exec/system_context.hpp
Comment (ericniebler) [inline include/exec/system_context.hpp:228]: why are we heap-allocating this? why can't it be part of the operation state?

File: include/exec/system_context.hpp
Comment (ericniebler) [inline include/exec/system_context.hpp:236]: so there are going to be `shape` indirect function calls? i'm curious if there's a better way.

File: include/exec/system_context.hpp
Comment (ericniebler) [inline include/exec/system_context.hpp:323]: This isn't correct. The `bulk` sender should send the values and errors of its predecessor.

File: include/exec/system_context.hpp
Comment (ericniebler) [reply include/exec/system_context.hpp:228]: Also, we should be storing the decayed values in the tuple.

Comment (lucteo) [reply include/exec/system_context.hpp:228]: We shouldn't allocate here; we should be able to tell size of all the possible `set_value` arguments when we build this object. This is on my list of TODOs, but didn't get to it yet.

On the decayed values, I didn't noticed that. Thank you for pointing it out.

File: include/exec/system_context.hpp
Comment (lucteo) [reply include/exec/system_context.hpp:236]: I doubt there is a better way. In the common case, each call will be made on a different thread.

In general, to schedule work, we need 1 + N calls across the external interface (possible to a different library), where N is the number of threads we need.

File: include/exec/system_context.hpp
Comment (lucteo) [reply include/exec/system_context.hpp:323]: Indeed, see Lee's TODO 2 lines below.

However, my understanding that this sender will always complete with `set_value()` with no arguments, regardless of the values sent by the previous senders; these values are just used to pass in to the given user function. Is my understanding wrong?

On the error side, my plan is to forward all errors given by the previous sender, plus `std::exception_ptr`, so that failure to schedule on the external scheduler can be reported as an error.

Comment (ericniebler) [reply include/exec/system_context.hpp:323]: > However, my understanding that this sender will always complete with `set_value()` with no arguments, regardless of the values sent by the previous senders

Not so. `bulk` is specified to move the stored values into the value channel after the bulk work is done.

**Comment by lucteo** [comment]
Dear race conditions and heisenbugs. I get it that you are angry with me, but now you really pissed me off. I've tuned already my profiler on you. Have a happy very-short life!

**Comment by lucteo** [comment]
> Dear race conditions and heisenbugs. I get it that you are angry with me, but now you really pissed me off. I've tuned already my profiler on you. Have a happy very-short life!

Found ya!
image

**Comment by lucteo** [comment]
@ericniebler : can you please give an /ok to test ?
(I took the latest bugs to be a personal insult ;) )

**Review by ericniebler — APPROVED** [review]

**Comment by ericniebler** [comment]
@lucteo are you planning any more work on this PR?

**Pull request merged**
